go/pkg/tool/tool.go

```diff
@@ -274,18 +273,47 @@ func (c *Client) DownloadBlob(ctx context.Context, blobDigest, path string) (str
 }
 
 // UploadBlob uploads a blob from the specified path into the remote cache.
-func (c *Client) UploadBlob(ctx context.Context, path string) error {
-	dg, err := digest.NewFromFile(path)
-	if err != nil {
-		return err
+func (c *Client) UploadBlob(ctx context.Context, paths ...string) error {
+	return c.UploadBlobs(ctx, 1, paths)
+}
+
+// UploadBlobs uploads blobs from the specified paths into the remote cache.
+func (c *Client) UploadBlobs(ctx context.Context, concurrency uint64, paths []string) error {
+	uploader := newParallelUploader(ctx, c.GrpcClient, concurrency)
+	for _, path := range paths {
+		if err := uploader.Enqueue(path); err != nil {
+			return err
+		}
 	}
+	return uploader.CloseAndWait()
+}
 
-	log.Infof("Uploading blob of %v from %v.", dg, path)
-	ue := uploadinfo.EntryFromFile(dg, path)
-	if _, _, err := c.GrpcClient.UploadIfMissing(ctx, ue); err != nil {
+// UploadTree uploads a tree from the specified path into the remote cache.
+func (c *Client) UploadTree(ctx context.Context, concurrency uint64, path string) error {
+	uploader := newParallelUploader(ctx, c.GrpcClient, concurrency)
```

Comment:
I wonder why couldn't you use `UploadIfMissing` directly instead of making a parallel uploaded? `UploadIfMissing` should already has [parallel support](https://github.com/bazelbuild/remote-apis-sdks/blob/fdb5dae38ca890ae7311dc0d320e2d327df10485/go/pkg/client/cas.go#L419), controlled by [CASConcurrency](https://github.com/bazelbuild/remote-apis-sdks/blob/fdb5dae38ca890ae7311dc0d320e2d327df10485/go/pkg/flags/flags.go#L46)

Reply:
+1 to this.

Reply:
I'm using this tool to upload very large trees with 10,000s files. Just iterating over all of them takes significant time, so it's more efficient to start uploading while iterating over the files in the tree.

I'll have a look if I can achieve this with the existing parallel uploading.